# **RNN**
A recurrent neural network (RNN) is a class of artificial neural network where connections between units form a directed cycle. This creates an internal state of the network which allows it to exhibit dynamic temporal behavior.

IMDB sentiment classification task

This is a dataset for binary sentiment classification containing substantially more data than previous benchmark datasets. IMDB provided a set of 25,000 highly polar movie reviews for training, and 25,000 for testing. There is additional unlabeled data for use as well. Raw text and already processed bag of words formats are provided.

You can download the dataset from http://ai.stanford.edu/~amaas/data/sentiment/  or you can directly use 
" from keras.datasets import imdb " to import the dataset.

Few points to be noted:
Modules like SimpleRNN, LSTM, Activation layers, Dense layers, Dropout can be directly used from keras
For preprocessing, you can use required 

In [1]:
#load the imdb dataset 
import tensorflow as tf
from keras.datasets import imdb

vocabulary_size = 5000
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words = vocabulary_size)
print('Loaded dataset with {} training samples, {} test samples'.format(len(X_train), len(X_test)))

/home/vishal/anaconda3/envs/smai/lib/python3.8/site-packages/tensorflow/python/keras/datasets/imdb.py:155: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/home/vishal/anaconda3/envs/smai/lib/python3.8/site-packages/tensorflow/python/keras/datasets/imdb.py:156: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


Loaded dataset with 25000 training samples, 25000 test samples


In [2]:
X_train[:5]

array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 2, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 2, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 2, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 2, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 2, 19, 178, 32]),
       list([1, 194, 1153, 194, 2, 78, 228, 5, 6, 1463, 4369,

In [3]:
#the review is stored as a sequence of integers. 
# These are word IDs that have been pre-assigned to individual words, and the label is an integer

# print('---review---')
# print(X_train[:])
# print('---label---')
# print(y_train[:])

# to get the actual review
word2id = imdb.get_word_index()
id2word = {i: word for word, i in word2id.items()}
# print('---review with words---')
# print([id2word.get(i, ' ') for i in X_train])
# print('---label---')
# print(y_train[:])
print(X_train.shape)
print(X_train[1])

(25000,)
[1, 194, 1153, 194, 2, 78, 228, 5, 6, 1463, 4369, 2, 134, 26, 4, 715, 8, 118, 1634, 14, 394, 20, 13, 119, 954, 189, 102, 5, 207, 110, 3103, 21, 14, 69, 188, 8, 30, 23, 7, 4, 249, 126, 93, 4, 114, 9, 2300, 1523, 5, 647, 4, 116, 9, 35, 2, 4, 229, 9, 340, 1322, 4, 118, 9, 4, 130, 4901, 19, 4, 1002, 5, 89, 29, 952, 46, 37, 4, 455, 9, 45, 43, 38, 1543, 1905, 398, 4, 1649, 26, 2, 5, 163, 11, 3215, 2, 4, 1153, 9, 194, 775, 7, 2, 2, 349, 2637, 148, 605, 2, 2, 15, 123, 125, 68, 2, 2, 15, 349, 165, 4362, 98, 5, 4, 228, 9, 43, 2, 1157, 15, 299, 120, 5, 120, 174, 11, 220, 175, 136, 50, 9, 4373, 228, 2, 5, 2, 656, 245, 2350, 5, 4, 2, 131, 152, 491, 18, 2, 32, 2, 1212, 14, 9, 6, 371, 78, 22, 625, 64, 1382, 9, 8, 168, 145, 23, 4, 1690, 15, 16, 4, 1355, 5, 28, 6, 52, 154, 462, 33, 89, 78, 285, 16, 145, 95]


In [12]:
set(y_train)

{0, 1}

In [4]:
#pad sequences (write your code here)
from keras.preprocessing import sequence

X_train = tf.keras.preprocessing.sequence.pad_sequences(X_train)
X_test = tf.keras.preprocessing.sequence.pad_sequences(X_test)

In [7]:
#design a RNN model (write your code)

from keras import Sequential
from keras.layers import Embedding, LSTM, Dense, Dropout, SimpleRNN

model = Sequential()
model.add(
    Embedding(
        vocabulary_size,
        64
    )
)
model.add(
   SimpleRNN(64) 
)
model.add(
    Dense(10)
)
model.add(
    Dropout(0.2)
)
model.add(
    Dense(1)
)


In [8]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 64)          320000    
_________________________________________________________________
simple_rnn (SimpleRNN)       (None, 64)                8256      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                650       
_________________________________________________________________
dropout_1 (Dropout)          (None, 10)                0         
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 11        
Total params: 328,917
Trainable params: 328,917
Non-trainable params: 0
_________________________________________________________________


In [9]:
#train and evaluate your model
#choose your loss function and optimizer and mention the reason to choose that particular loss function and optimizer
# use accuracy as the evaluation metric

model.compile(
    loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
    optimizer=tf.keras.optimizers.Adam(),
    metrics=['accuracy']
)

# Reason for choosing
1. ADAM is currently one of the best optimizers in the market. It combines the advantages of both rms and ADAGRAD, making it more robust towards **Vanishing and exploding gradients**. Main advantages that I found useful include:
    1. It handles deep and complex networks like RNN and LSTMs very well (as it suppresses the problem of vanishing and expoding gradients)
    2. It does not oscillate near local and global minima due to the momentum factor
    3. Adaptive learning rate reduces the requirement of hypertuning
2. Binary cross entropy is chosen because here we are trying to perform binary classification here. This function works well for multi-class classification as well (if the output is one-hot type).

In [15]:
batch_size = 64
num_epochs = 2

model.fit(
    X_train, 
    y_train,
    epochs=num_epochs,
    batch_size=batch_size
)

2021-12-04 17:58:51.381640: W tensorflow/core/framework/cpu_allocator_impl.cc:81] Allocation of 249400000 exceeds 10% of free system memory.


Epoch 1/2


2021-12-04 17:58:52.284092: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10


391/391 [==============================] - 601s 2s/step - loss: 0.5550 - accuracy: 0.6756
Epoch 2/2
391/391 [==============================] - 883s 2s/step - loss: 0.3448 - accuracy: 0.8520


In [16]:
#evaluate the model using model.evaluate()
test_loss, test_acc = model.evaluate(X_test[:1000], y_test[:1000], batch_size=64)

16/16 [==============================] - 2s 129ms/step - loss: 0.4182 - accuracy: 0.8250


In [17]:
print('Test Loss: {}\nTest Accuracy: {}'.format(test_loss, test_acc))

Test Loss: 0.4181668162345886
Test Accuracy: 0.824999988079071


# **LSTM**

Instead of using a RNN, now try using a LSTM model and compare both of them. Which of those performed better and why ?


In [18]:
model = Sequential()
model.add(
    Embedding(
        vocabulary_size,
        64
    )
)
model.add(
    LSTM(64)
)
model.add(
    Dense(10)
)
model.add(
    Dropout(0.2)
)
model.add(
    Dense(1)
)

In [23]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, None, 64)          320000    
_________________________________________________________________
lstm_1 (LSTM)                (None, 64)                33024     
_________________________________________________________________
dense_4 (Dense)              (None, 10)                650       
_________________________________________________________________
dropout_2 (Dropout)          (None, 10)                0         
_________________________________________________________________
dense_5 (Dense)              (None, 1)                 11        
Total params: 353,685
Trainable params: 353,685
Non-trainable params: 0
_________________________________________________________________


In [19]:
model.compile(
    loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
    optimizer=tf.keras.optimizers.Adam(),
    metrics=['accuracy']
)

In [20]:
model.fit(
    X_train, 
    y_train,
    epochs=2,
    batch_size=64
)

2021-12-04 18:26:59.350418: W tensorflow/core/framework/cpu_allocator_impl.cc:81] Allocation of 249400000 exceeds 10% of free system memory.


Epoch 1/2


2021-12-04 18:27:00.989741: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7


391/391 [==============================] - 53s 135ms/step - loss: 0.4381 - accuracy: 0.7709
Epoch 2/2
391/391 [==============================] - 52s 134ms/step - loss: 0.2818 - accuracy: 0.8853


In [21]:
test_loss, test_acc = model.evaluate(X_test[:1000], y_test[:1000], batch_size=64)

16/16 [==============================] - 1s 47ms/step - loss: 0.2987 - accuracy: 0.8580


In [22]:
print('Test Loss: {}\nTest Accuracy: {}'.format(test_loss, test_acc))

Test Loss: 0.2986765503883362
Test Accuracy: 0.8579999804496765


# Observations (LSTM vs RNN)
1. RNNs took too much time to train when compared to LSTMs although LSTMs had more number of parameters than RNN. This is a bit controversial as RNNs are less complex and usually take less time to train. This could have happened because the conditions were not constant and perfect enough to benchmark their performances (as PCs are not ideal for benchmarking). 
2. LSTMs gave better train and test accuracies than RNNs 
    - LSTM accuracy on test set: 85.8%
    - RNN accuracy on test set: 82.50%

Perform Error analysis and explain using few examples.